facilitator/src/http.rs

```diff
@@ -25,23 +26,112 @@ impl Method {
+    additional_retryable_http_codes: Option<Vec<u16>>,
 }
 
-/// Prepares a request by taking in an agent and the RequestParameters. Could
-/// return an Error if the OauthTokenProvider returns an error when supplying
-/// the request with an Oauth token.
-pub(crate) fn prepare_request(agent: &Agent, parameters: RequestParameters<'_>) -> Result<Request> {
-    let request = agent.request_url(parameters.method.to_primitive_string(), &parameters.url);
-    parameters.authenticated_request(request)
+impl Default for RetryingAgent {
+    fn default() -> Self {
+        Self::new(
+            AgentBuilder::new().timeout(Duration::from_secs(10)).build(),
+            None,
+        )
+    }
 }
 
-/// Prepares a request and wraps it with an agent that has a timeout of 10
-/// seconds. Read prepare_request for more information.
-pub(crate) fn prepare_request_without_agent(parameters: RequestParameters<'_>) -> Result<Request> {
-    prepare_request(&create_agent(), parameters)
+impl RetryingAgent {
+    pub fn new(agent: Agent, additional_retryable_http_codes: Option<Vec<u16>>) -> Self {
+        Self {
+            agent,
+            additional_retryable_http_codes,
+        }
+    }
+
+    /// Prepares a request for the provided `RequestParameters`. Returns a
+    /// `ureq::Request` permitting the caller to further customize the request
+    /// (e.g., with HTTP headers or query parameters). Callers may use methods
+    /// like `send()` or `send_bytes()` directly on the returned `Request`, but
+    /// must use `RetryingAgent::send_json_request`, `::send_bytes` or
+    /// `::send_string` to get retries.
+    /// Returns an Error if the OauthTokenProvider returns an error when
+    /// supplying the request with an OauthToken.
+    pub(crate) fn prepare_request(&self, parameters: RequestParameters) -> Result<Request> {
+        let mut request = self
+            .agent
+            .request_url(parameters.method.to_primitive_string(), &parameters.url);
+        if let Some(token_provider) = parameters.token_provider {
+            let token = token_provider.ensure_oauth_token()?;
+            request = request.set("Authorization", &format!("Bearer {}", token));
+        }
+        Ok(request)
+    }
+
+    fn is_http_status_retryable(&self, http_status: u16) -> bool {
+        if http_status >= 500 {
+            return true;
+        }
+
+        if let Some(codes) = &self.additional_retryable_http_codes {
+            return codes.contains(&http_status);
+        }
+
+        false
+    }
+
+    fn is_error_retryable(&self, error: &ureq::Error) -> bool {
+        match error {
+            ureq::Error::Status(http_status, _) => self.is_http_status_retryable(*http_status),
+            ureq::Error::Transport(_) => true,
+        }
+    }
+
+    /// Send the provided request with the provided JSON body.
+    pub(crate) fn send_json_request(&self, request: Request, body: SerdeValue) -> Result<Response> {
```

Comment:
If you're cloning Request and SerdeValue anyhow, why take ownership? I think these should be `&`

Reply:
I was somewhat blindly replicating the API of the underlying `ureq::Request` methods, but you are right, these can be references.